In [1]:
import numpy as np   #Importing Numpy Module 

#### Automatic generation of Numbers in array

In [2]:
a = np.arange(10) #Default 0 is the starting value and stop value is given as 10 in the example
a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [4]:
a= np.arange(10,20,2) #Start=10 End =20 (Excluded) , Jump=2
a

array([10, 12, 14, 16, 18])

#### Why Numpy is better than lists?

In [5]:
L = range(1000) #Created a List with 1000 elements from 0 to 999
%timeit [i**2 for i in L] #timeit is magical function that is available in Jupyter notebook

832 µs ± 34.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


So we got 832µs per loop while doing multiplication by 2 for each element in list of 100 elements

In [6]:
L = np.arange(1000) #Created a Numpy array with 1000 elements from 0 to 999
%timeit L**2

3.61 µs ± 108 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


It just took 3.61µs per loop which is around 250 times better

#### Creating A Array

In [7]:
a= np.array([3,5,6,1,7]) #Creating a Numpy array with List elements
a

array([3, 5, 6, 1, 7])

In [8]:
a.ndim #Will return the Number of Dimension (It is 1 dimension and it is scalar)

1

In [9]:
a.shape #Since it is only 1 dimension and it returns the number of elements in that array, will understand its use better in 2D array

(5,)

In [10]:
len(a) #Will return the number of elements in the list

5

#### Create a 2D array

In [16]:
a = np.array([[2,3,5],[3,4,5]]) # Creating a 2D Array/Matrix # It is like List of list
a

array([[2, 3, 5],
       [3, 4, 5]])

In [17]:
type(a) #Checking the datatype of this array - n dimensional array

numpy.ndarray

In [18]:
a.ndim #It is 2 dimension array

2

In [19]:
a.shape #For 2 dimension arrays shape will return rows,columns , Here rows=2, columns=3

(2, 3)

In [20]:
len(a) #It will return the number of rows

2

#### Create a 3D array

In [41]:
a = np.array([            #Creating 3 Dimension array
    [[0,1,2],[1,2,3]],
    [[3,4,5],[5,6,7]]
    ])
a

array([[[0, 1, 2],
        [1, 2, 3]],

       [[3, 4, 5],
        [5, 6, 7]]])

In [42]:
a.ndim 

3

In [44]:
a.shape #It is of 2 breadth (2), then rows,columns

(2, 2, 3)

#### Functions for creating arrays

In [45]:
a= np.arange(2,15,3) #start, end(excluded),step-Index
a

array([ 2,  5,  8, 11, 14])

In [47]:
# Using linspace

b=np.linspace(1,3,6) #Return 6 points in between 1 and 3

b

array([1. , 1.4, 1.8, 2.2, 2.6, 3. ])

In [49]:
c= np.ones((2,2)) # Creates a 2 * 2 matrix  with all values as 1
c

array([[1., 1.],
       [1., 1.]])

In [50]:
d= np.zeros((2,2)) # Creates a 2 * 2 matrix  with all values as 0
d

array([[0., 0.],
       [0., 0.]])

In [52]:
e=np.eye(3) #All the diagonal numbers are filled with value 1. It is unit matrix or Identity Matrix
e

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [54]:
f=np.eye(3,2) # 1*1, 2*2, 3*3 will have value 1 for Identity Matrix
f

array([[1., 0.],
       [0., 1.],
       [0., 0.]])

In [56]:
#create Array Using diagonal function

a= np.diag([1,2,4,5])

a

array([[1, 0, 0, 0],
       [0, 2, 0, 0],
       [0, 0, 4, 0],
       [0, 0, 0, 5]])

In [57]:
#extracting Diagonal Elements

np.diag(a)

array([1, 2, 4, 5])

In [58]:
#Create Array with random numbers

a=np.random.rand(4) #Creates a random array with elements between 0 and 1

a #These are uniform

array([0.94937761, 0.92391463, 0.93707447, 0.62910368])

In [62]:
a= np.random.randn(4)

a # These are probability and statistics

array([-0.84917265, -1.54297681, -0.16968304,  1.93234239])

#### Datatypes

In [64]:
a = np.arange(10)

print(a)

print(a.dtype)

[0 1 2 3 4 5 6 7 8 9]
int32


In [66]:
#We can change the dtype explicitly

a=np.arange(10,dtype=float)

print(a)

print(a.dtype)

[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
float64


In [67]:
np.ones((3,3)) # will always create with float numbers

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [68]:
np.ones((3,3),dtype='int64')

array([[1, 1, 1],
       [1, 1, 1],
       [1, 1, 1]], dtype=int64)

In [70]:
#Real and Imaginary Numbers

a = np.array([1+2j,3+5j])

a.dtype

dtype('complex128')

In [72]:
#Create Boolean Numpy array

b= np.array([True,False,False])

b.dtype

dtype('bool')

In [73]:
#Create String Numpy Array

c = np.array(['Santhosh','Vijay'])

c.dtype

dtype('<U8')

#### Indexing and Slicing

##### Indexing

In [75]:
a = np.arange(10)
print(a)
a[3] # Will return the value with Index 3 . Index starts from 0

[0 1 2 3 4 5 6 7 8 9]


3

In [81]:
a= np.diag([1,2,3,5])

print(a)

a[1][1] # will return the 2nd row and 2 column intersection value

[[1 0 0 0]
 [0 2 0 0]
 [0 0 3 0]
 [0 0 0 5]]


2

In [82]:
#We can assing the value also

a[1][1]=20

a #Now the value 2 is replaced with 20

array([[ 1,  0,  0,  0],
       [ 0, 20,  0,  0],
       [ 0,  0,  3,  0],
       [ 0,  0,  0,  5]])

##### Slicing

In [83]:
a = np.arange(10)
a[2:5] #Will return elements from Index 2 to 4

array([2, 3, 4])

In [84]:
a[2:8:2] #Returns the element from Index 2 to 7 with Step Index of 2

array([2, 4, 6])

In [85]:
a[::-1] #Returns all the element with step index of -1 i.e Reverse

array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

In [86]:
#We can also assign values to multiple elements

a[3:6] = 50 #Will assign the value 50 to Elements from Index 3 to 5

a

array([ 0,  1,  2, 50, 50, 50,  6,  7,  8,  9])

In [87]:
b = np.arange(5)

a[5:]=b[::-1] # So Reversing the b array and assign to a array from index 5 to last

a

array([ 0,  1,  2, 50, 50,  4,  3,  2,  1,  0])

So you can see the last 5 values are  4,  3,  2,  1,  0 which is reverse of array b

#### Copies and Views

In [89]:
a=np.arange(10)

a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [90]:
b= a[0:5] ## B array will hold the values of index from 0 to 4

b

array([0, 1, 2, 3, 4])

In [91]:
#Now modifying b array

b[0]=100

b

array([100,   1,   2,   3,   4])

In [92]:
#Now lets check a array

a #Here a[0] is changed to value 100

array([100,   1,   2,   3,   4,   5,   6,   7,   8,   9])

Numpy produces a View and not a copy when you are storing a slicing value to another array.
Numpy does that for memory efficiency.

In [93]:
np.shares_memory(a,b) #Will check whether both array saves in a same place

True

In [94]:
#Since a and b are pointing to the same location, I will modify values of a to checker whether it is reflecting in b

a[1]=50

b

array([100,  50,   2,   3,   4])

Yes it gets modified

In [95]:
#how to avoid this
#We can use copy explicity to make a copy and not a view

c = a[0:5].copy()

c

array([100,  50,   2,   3,   4])

In [96]:
a[0]=250

c #It doesn't reflected

array([100,  50,   2,   3,   4])

In [97]:
np.shares_memory(a,c)

False

#### Fancy Indexing

In [115]:
a= np.random.randint(0,20,15) #Will return 15 numbers between values 0 and 15
a

array([10,  3,  5,  3, 11,  6, 10,  8, 19,  3,  5, 15,  8,  6, 12])

In [116]:
#For Example, I want to retrive all the even elements in the array

a%2==0  #Will return whether each elements is odd or Even . Here True represents the number is even


array([ True, False, False, False, False,  True,  True,  True, False,
       False, False, False,  True,  True,  True])

In [117]:
a[a%2==0] #This will return the even numbers in the array

array([10,  6, 10,  8,  8,  6, 12])

In [118]:
a[a%2==0]=1 #It will modify all the even values to number 1

In [119]:
a

array([ 1,  3,  5,  3, 11,  1,  1,  1, 19,  3,  5, 15,  1,  1,  1])

In [120]:
a=np.arange(0,100,10)
a

array([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [122]:
a[[2,3,6,7]] #Will return 2,3,6 and 7 th index element

#When giving multiple indexing please make sure in giving in list of list

array([20, 30, 60, 70])

In [124]:
a[[7,8]]=400 #Modifying 7th and 8th index element to 400
a

array([  0,  10,  20,  30,  40,  50,  60, 400, 400,  90])